File: jeffrey-microscope/profiles/timeseries/src/test/java/cafe/jeffrey/timeseries/TimeseriesDownsamplerTest.java
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cafe.jeffrey.timeseries;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class TimeseriesDownsamplerTest {

    private static SingleSerie serie(String name, int seconds) {
        List<List<Long>> data = new ArrayList<>();
        for (long second = 0; second < seconds; second++) {
            List<Long> point = new ArrayList<>(2);
            point.add(second);
            point.add(1L);
            data.add(point);
        }
        return new SingleSerie(name, data);
    }

    private static long totalValue(SingleSerie serie) {
        return serie.data().stream().mapToLong(point -> point.get(1)).sum();
    }

    private static SingleSerie byName(TimeseriesData data, String name) {
        return data.series().stream()
                .filter(serie -> serie.name().equals(name))
                .findFirst()
                .orElseThrow();
    }

    @Nested
    class SingleSerieInput {

        @Test
        void boundsThePointCount() {
            TimeseriesData input = new TimeseriesData(serie("Samples", 10_000));

            TimeseriesData result = TimeseriesDownsampler.downsample(input, 1_000);

            assertEquals(1, result.series().size());
            assertTrue(byName(result, "Samples").data().size() <= 1_000);
        }

        @Test
        void sumPreservesTotalValue() {
            TimeseriesData input = new TimeseriesData(serie("Samples", 10_000));

            TimeseriesData result = TimeseriesDownsampler.downsample(input, 1_000);

            assertEquals(10_000L, totalValue(byName(result, "Samples")));
        }

        @Test
        void aggregatesSpikeMassIntoItsBucket() {
            List<List<Long>> data = new ArrayList<>();
            for (long second = 0; second < 1_000; second++) {
                List<Long> point = new ArrayList<>(2);
                point.add(second);
                point.add(second == 500 ? 400L : 1L);
                data.add(point);
            }
            TimeseriesData input = new TimeseriesData(new SingleSerie("Samples", data));

            TimeseriesData result = TimeseriesDownsampler.downsample(input, 100);

            // Total mass (999 ones + one 400) is preserved by SUM.
            assertEquals(999L + 400L, totalValue(byName(result, "Samples")));
        }

        @Test
        void smallInputIsReturnedUnchanged() {
            TimeseriesData input = new TimeseriesData(serie("Samples", 50));

            TimeseriesData result = TimeseriesDownsampler.downsample(input, 1_000);

            assertEquals(1, result.series().size());
            assertEquals(50, byName(result, "Samples").data().size());
        }
    }

    @Nested
    class MultiSerieInput {

        @Test
        void reducesEachSerieKeepingSeriesCount() {
            TimeseriesData input = new TimeseriesData(
                    serie("Samples", 5_000),
                    serie("Matched Samples", 5_000));

            TimeseriesData result = TimeseriesDownsampler.downsample(input, 500);

            assertEquals(2, result.series().size());
            assertTrue(byName(result, "Samples").data().size() <= 500);
            assertTrue(byName(result, "Matched Samples").data().size() <= 500);
        }
    }
}

File: jeffrey-microscope/pages-microscope/src/services/timeseries/TimeseriesDownsampler.test.ts
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { describe, expect, it } from 'vitest';
import TimeseriesDownsampler from '@/services/timeseries/TimeseriesDownsampler';

const zeroSeededSeconds = (length: number): number[][] => {
  const data: number[][] = [];
  for (let second = 0; second < length; second++) {
    data.push([second, 0]);
  }
  return data;
};

describe('TimeseriesDownsampler', () => {
  describe('downsamplePeaks', () => {
    it('returns the input untouched when it fits under the cap', () => {
      const data = zeroSeededSeconds(100);
      expect(TimeseriesDownsampler.downsamplePeaks(data, 5000)).toBe(data);
    });

    it('returns the input untouched when it sits exactly on the cap', () => {
      const data = zeroSeededSeconds(1500);
      expect(TimeseriesDownsampler.downsamplePeaks(data, 1500)).toBe(data);
    });

    it('returns an empty series unchanged', () => {
      expect(TimeseriesDownsampler.downsamplePeaks([], 1500)).toEqual([]);
    });

    it('treats a non-positive cap as a no-op', () => {
      const data = zeroSeededSeconds(10);
      expect(TimeseriesDownsampler.downsamplePeaks(data, 0)).toBe(data);
      expect(TimeseriesDownsampler.downsamplePeaks(data, -1)).toBe(data);
    });

    it('keeps isolated spikes that every-nth-point sampling would drop', () => {
      // A 10h recording seeded per second, with spikes on seconds that are deliberately not
      // multiples of the old sampler's step (ceil(36022 / 1500) = 25) — every one of them was
      // invisible before. This is the VM Operations safepoint-pause case.
      const spikeSeconds = [7, 33, 1111, 20004, 36017];
      const spikeValue = 5_000_000;
      const data = zeroSeededSeconds(36022);
      for (const second of spikeSeconds) {
        data[second][1] = spikeValue;
      }

      const result = TimeseriesDownsampler.downsamplePeaks(data, 1500);

      expect(result.length).toBeLessThanOrEqual(1500);
      const nonZero = result.filter(point => point[1] !== 0);
      expect(nonZero).toHaveLength(spikeSeconds.length);
      expect(Math.max(...result.map(point => point[1]))).toBe(spikeValue);

      // Each surviving spike still sits within one bucket of the second it actually happened.
      const bucketSize = Math.ceil(data.length / 1500);
      nonZero.forEach((point, index) => {
        const trueSecond = spikeSeconds[index];
        expect(point[0]).toBeLessThanOrEqual(trueSecond);
        expect(trueSecond - point[0]).toBeLessThan(bucketSize);
      });
    });

    it('emits the peak of every bucket at the x of the bucket start', () => {
      const data = [];
      for (let i = 0; i < 20; i++) {
        data.push([i, i]);
      }

      expect(TimeseriesDownsampler.downsamplePeaks(data, 5)).toEqual([
        [0, 3],
        [4, 7],
        [8, 11],
        [12, 15],
        [16, 19]
      ]);
    });

    it('decimates every series onto the same x grid regardless of its values', () => {
      const flat = zeroSeededSeconds(1000);
      const ramp = zeroSeededSeconds(1000).map(point => [point[0], point[0]]);
      const spiky = zeroSeededSeconds(1000);
      [3, 17, 402, 913].forEach(second => {
        spiky[second][1] = 999;
      });

      const xGrid = (data: number[][]): number[] => {
        return TimeseriesDownsampler.downsamplePeaks(data, 100).map(point => point[0]);
      };

      expect(xGrid(ramp)).toEqual(xGrid(flat));
      expect(xGrid(spiky)).toEqual(xGrid(flat));
    });

    it('handles a trailing partial bucket without reading past the end', () => {
      const data = [];
      for (let i = 0; i < 10; i++) {
        data.push([i, i]);
      }

      // bucketSize = ceil(10 / 3) = 4, so the last bucket covers indices 8-9 only.
      expect(TimeseriesDownsampler.downsamplePeaks(data, 3)).toEqual([
        [0, 3],
        [4, 7],
        [8, 9]
      ]);
    });
  });
});
